Add support for NXP S32K148 evaluation board

marcin-wierzbicki opened the pull request:
Add support for NXP S32K148 evaluation board.

The board can be debugged using "west debug" command (see boards/nxp/s32k148_evb/doc/index.rst).

Adapt samples: adc_dt, adc_sequence.
Adapt tests: adc_api, gpio_basic_api, gpio_hogs.

See [Add support for S32K148 SoC](https://github.com/zephyrproject-rtos/hal_nxp/pull/509)

maass-hamburg commented on drivers/ethernet/nxp_enet/Kconfig:
```diff
@@ -88,6 +88,12 @@ config ETH_NXP_ENET_RX_THREAD_PRIORITY
 	  ENET MAC Driver handles RX in cooperative workqueue thread.
 	  This options sets the priority of that thread.
 
+config ETH_NXP_ENET_IGNORE_RX_CRC_PHY_ERROR
+	bool "Enable ENET workaround for CRC Phy error"
+	default n
+	help
+	  Enable this option to apply workarounds for known ENET CRC issues on S32K148.
+
```
Where is that used? Can't find it

marcin-wierzbicki replied:
Please check this PR#509 in hal_nxp:
https://github.com/zephyrproject-rtos/hal_nxp/pull/509/commits/3e9922d516c119cbacd476d962ad5fb28101bf3f

manuargue replied:
The fix should be submitted to https://github.com/nxp-mcuxpresso/mcux-sdk. Because is a feature of the baremetal driver, it should use `FSL_FEATURE_ENET_` macros to conditionally compile code and not Kconfig symbols from Zephyr

marcin-wierzbicki replied:
@KevShaju: please take care of that.

KevShaju replied:
Hi @manuargue,
The [https://github.com/nxp-mcuxpresso/mcux-sdk](url) repository is set to be archived. Is it still appropriate to upstream to this repository?

manuargue replied:
yes, please submit it to that repo

manuargue commented:
>Add driver for TJA1101 PHY device.

I'd recommend to keep this first PR smaller including basic enablement of the SoC and board, and split the PHY driver into another PR for easier review. This will likely also help you to move the prs faster.

nordicjm commented on boards/nxp/s32k148_evb/doc/img/s32k148_evb.webp:
remove whitespace (ideally make background transparent) then trim image then vastly reduce resolution

nordicjm commented on boards/nxp/s32k148_evb/doc/index.rst:
```diff
@@ -0,0 +1,190 @@
+Supported Features
+==================
+
+The ``s32k148_evb`` board configuration supports the following hardware features:
```
```suggestion
The ``s32k148_evb`` board supports the following hardware features:
```

nordicjm replied:
stop marking things as resolved if they are not resolved and fix the highlighted issue

nordicjm commented on boards/nxp/s32k148_evb/doc/index.rst:
```diff
@@ -0,0 +1,190 @@
+
+The ``s32k148_evb`` board configuration supports the following hardware features:
+
+============  ==========  ================================
+Interface     Controller  Driver/Component
+============  ==========  ================================
+SYSMPU        on-chip     mpu
```
this doesn't look like how tables are done

nordicjm commented on boards/nxp/s32k148_evb/s32k148_evb.dts:
```diff
@@ -0,0 +1,180 @@
+			gpios = <&gpioe 21 GPIO_ACTIVE_HIGH>;
+			label = "LED1_RGB_RED";
+		};
+		led1_green: led_1 {
```
needs spacing

nordicjm commented on boards/nxp/s32k148_evb/s32k148_evb_defconfig:
```diff
@@ -0,0 +1,14 @@
+# Enable MPU
+CONFIG_ARM_MPU=y
+
+CONFIG_PINCTRL=y
```
not allowed

nordicjm commented on drivers/ethernet/phy/CMakeLists.txt:
```diff
@@ -3,6 +3,7 @@
 zephyr_library_sources_ifdef(CONFIG_PHY_GENERIC_MII	phy_mii.c)
 zephyr_library_sources_ifdef(CONFIG_PHY_ADIN2111	phy_adin2111.c)
 zephyr_library_sources_ifdef(CONFIG_PHY_DM8806	phy_dm8806.c)
+zephyr_library_sources_ifdef(CONFIG_PHY_TJA1101	        phy_tja1101.c)
```
```suggestion
zephyr_library_sources_ifdef(CONFIG_PHY_TJA1101	phy_tja1101.c)
```

marcin-wierzbicki replied:
@KevShaju please address this

marcin-wierzbicki replied:
will be handled in a separate PR

KevShaju replied:
Fixed in commit `4a8e045f07c73407d2039735da1aed0eaafaf13a`

nordicjm commented on drivers/ethernet/nxp_enet/Kconfig:
```diff
@@ -88,6 +88,12 @@ config ETH_NXP_ENET_RX_THREAD_PRIORITY
 
+config ETH_NXP_ENET_IGNORE_RX_CRC_PHY_ERROR
+	bool "Enable ENET workaround for CRC Phy error"
+	default n
```
```suggestion
```

marcin-wierzbicki replied:
@KevShaju please address this

marcin-wierzbicki replied:
will be handled in a separate PR

KevShaju replied:
Removed this Kconfig option.

nordicjm commented on dts/arm/nxp/nxp_s32k148.dtsi:
```diff
@@ -30,6 +30,30 @@
+			compatible = "nxp,enet";
+			reg = <0x40079000 0x628>;
+			clocks = <&clock NXP_S32_ENET_CLK>;
+			enet_mac: ethernet {
```
lacks spacing

marcin-wierzbicki replied:
will be handled in a separate PR

nordicjm replied:
Err no, handle in this PR as per coding guidelines https://docs.zephyrproject.org/latest/contribute/style/devicetree.html

> Insert a single empty line to separate nodes at the same hierarchy level.

KevShaju replied:
Added spacing as per guidelines.

marcin-wierzbicki commented on boards/nxp/s32k148_evb/s32k148_evb.dts:
```diff
@@ -0,0 +1,201 @@
+	pinctrl-0 = <&pinmux_enet_mdio>;
+	pinctrl-names = "default";
+	phy: phy@0 {
+		compatible = "nxp,tja1101";
```
@KevShaju please take care of this

marcin-wierzbicki replied:
will be handled in a separate PR

KevShaju replied:
Added the PHY as a shield in `7e879ecb6fe5c40a045a737a11ba8ad4458bd453`.

marcin-wierzbicki commented on drivers/ethernet/phy/phy_tja1101.c:
```diff
@@ -0,0 +1,274 @@
+LOG_MODULE_REGISTER(phy_tja1101, CONFIG_PHY_LOG_LEVEL);
+
+/* Basic control register */
+#define TJA1101_BASIC_CONTROL              0x0000U
```
@KevShaju please address this

marcin-wierzbicki replied:
will be handled in a separate PR

KevShaju replied:
Updated based on your suggestion.

marcin-wierzbicki commented on drivers/ethernet/phy/phy_tja1101.c:
@KevShaju let's handle that in a separate PR

manuargue replied:
> and the existing tj1103 driver can be renamed later to c45_tja11xx

note that I'm not requesting you to do this change, it's out of the scope of your changes

> This new driver can serve as the first version of tja11xx

this is the actual request, to rename your tja1104.c to tja11xx.c and use it as the first version of the generic c22 tja11xx driver.

marcin-wierzbicki replied:
will be handled in a separate PR

KevShaju replied:
Renamed the driver to tja11xx.c as suggested.

marcin-wierzbicki commented:
> > Add driver for TJA1101 PHY device.
> 
> I'd recommend to keep this first PR smaller including basic enablement of the SoC and board, and split the PHY driver into another PR for easier review. This will likely also help you to move the prs faster. This is not mandatory, but in general a good practice.

Sure. Thank you.

marcin-wierzbicki commented on boards/nxp/s32k148_evb/s32k148_evb.yaml:
```diff
@@ -0,0 +1,22 @@
+  - zephyr
+supported:
+  - mpu
+  - gpio
```
Enabled gpio_basic_tests, gpio_hogs

nordicjm commented:
@marcin-wierzbicki you have marked multiple comments I left as resolved, they aren't resolved, you haven't fixed the issues, please acquaint yourself with the contributor guidelines https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html#workflow-suggestions-that-help-reviewers as per

> Unless they applied the reviewer’s recommendation exactly, authors must not resolve and hide comments, they must let the initial reviewer do it. The Zephyr project does not require all comments to be resolved before merge. Leaving some completed discussions open can sometimes be useful to understand the greater picture.

Also apply the coding style requirements to this PR without saying "will be handled in future"

maass-hamburg commented:
> > @marcin-wierzbicki you have marked multiple comments I left as resolved, they aren't resolved, you haven't fixed the issues, please acquaint yourself with the contributor guidelines https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html#workflow-suggestions-that-help-reviewers as per
> > > Unless they applied the reviewer’s recommendation exactly, authors must not resolve and hide comments, they must let the initial reviewer do it. The Zephyr project does not require all comments to be resolved before merge. Leaving some completed discussions open can sometimes be useful to understand the greater picture.
> > 
> > 
> > Also apply the coding style requirements to this PR without saying "will be handled in future"
> 
> Thank you for the feedback! The coding style fixes are already being addressed in a separate branch, and @KevShaju will be submitting a dedicated PR for them soon. This will help keep this PR focused on its main changes while ensuring consistency across the project.

This PR will not be approved and merged, if the requested changes are not done.

marcin-wierzbicki commented:
> > > @marcin-wierzbicki you have marked multiple comments I left as resolved, they aren't resolved, you haven't fixed the issues, please acquaint yourself with the contributor guidelines https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html#workflow-suggestions-that-help-reviewers as per
> > > > Unless they applied the reviewer’s recommendation exactly, authors must not resolve and hide comments, they must let the initial reviewer do it. The Zephyr project does not require all comments to be resolved before merge. Leaving some completed discussions open can sometimes be useful to understand the greater picture.
> > > 
> > > 
> > > Also apply the coding style requirements to this PR without saying "will be handled in future"
> > 
> > 
> > Thank you for the feedback! The coding style fixes are already being addressed in a separate branch, and @KevShaju will be submitting a dedicated PR for them soon. This will help keep this PR focused on its main changes while ensuring consistency across the project.
> 
> This PR will not be approved and merged, if the requested changes are not done.

Thank you. We have addressed all the requested changes in this PR.

marcin-wierzbicki commented on boards/nxp/s32k148_evb/doc/index.rst:
```diff
@@ -0,0 +1,200 @@
+===========================
+
+This board integrates an OpenSDA debug adapter. It can be used for flashing and debugging.
+The board cannot be debugged using the ``west debug`` command, since pyOCD does not support the target.
```
Thanks for the input. Regarding the nxp_s32dbg runner, I believe nxp_s32dbg.py won’t work for the S32K148 as-is. There are some hardcoded elements, and it currently runs the gta process, whereas in this case, we need to use pegdbserver_console. It might be necessary to introduce a new runner to support this properly

marcin-wierzbicki replied:
@manuargue Please review the implementation of the nxp_pedbg.py runner

Dat-NguyenDuy commented on dts/bindings/ethernet/nxp,tja11xx.yaml:
```diff
@@ -0,0 +1,29 @@
+# Copyright (c) 2023 NXP
```
Please remove `(c)` here

Dat-NguyenDuy reviewed (changes requested):
Some minor comments

nordicjm commented on boards/nxp/s32k148_evb/s32k148_evb-pinctrl.dtsi:
```diff
@@ -0,0 +1,86 @@
+			bias-pull-up;
+			slew-rate = "fast";
+		};
+		group1 {
```
no spacing has been added

nordicjm commented on boards/nxp/s32k148_evb/doc/index.rst:
```diff
@@ -0,0 +1,203 @@
+.. table:: RGB LED as GPIO LED
+   :widths: auto
+
+   ===============  ================  ===============  =====
+   Devicetree node  Devicetree alias  Label            Pin
+   ===============  ================  ===============  =====
+   led1_red         led0              LED1_RGB_RED     PTE21
+   led1_green       led1              LED1_RGB_GREEN   PTE22
+   led1_blue        led2              LED1_RGB_BLUE    PTE23
+   ===============  ================  ===============  =====
```
tables do not look valid e.g.
```
    +-------------------+-----------------------------------------------+
    | ``Command ID``    | Command description                           |
    +===================+===============================================+
    | ``0``             | Echo                                          |
    +-------------------+-----------------------------------------------+
```
but doc build fails to not able to tell

nordicjm commented on boards/shields/nxp_adtja1101/nxp_adtja1101.overlay:
```diff
@@ -0,0 +1,20 @@
+
+&enet_mdio {
+	status = "okay";
+	phy: phy@0 {
```
as above

nordicjm commented on tests/drivers/adc/adc_api/boards/s32k148_evb.overlay:
```diff
@@ -0,0 +1,43 @@
+		zephyr,resolution = <12>;
+	};
+
+	pinctrl-0 = <&adc0_default>;
+	pinctrl-names = "default";
```
move above child node

nordicjm commented on boards/nxp/s32k148_evb/doc/img/s32k148_evb.jpg:
convert to webp then put through https://tinypng.com/

maass-hamburg commented:
There is already a driver for the tja1103, maybe that driver can be extended to support all tja10xx phys?

KevShaju commented:
@maass-hamburg please see [comment](https://github.com/zephyrproject-rtos/zephyr/pull/85555#discussion_r1952816754). As suggested by @manuargue, we'll have two drivers phy_tja11xx.c and phy_c45_tja11xx.c similar to Linux, with this being the first version of tja11xx.

maass-hamburg commented on drivers/ethernet/phy/phy_tja11xx.c:
```diff
@@ -0,0 +1,269 @@
+	/* If Interrupt is configured then the workqueue will not
+	 * update the link state periodically so do it explicitly
+	 */
+	if (cfg->gpio_interrupt.port != NULL) {
```
I don't see the interrupt registered and so on, please add the functionality, so that it works, or remove it completely.

KevShaju replied:
removed the interrupts.

maass-hamburg commented on dts/bindings/ethernet/nxp,tja11xx.yaml:
```diff
@@ -0,0 +1,28 @@
+    required: true
+    description: PHY address
+
+  int-gpios:
+    type: phandle-array
+    description:
+      interrupt GPIO for PHY. Will be pulled high in its default state.
```
shouldn't be there, as long as the driver does not support it.

KevShaju replied:
removed.

maass-hamburg reviewed (changes requested):
also as said by @manuargue I would also prefer it, that the eth phy driver got it's own PR.

maass-hamburg commented on drivers/ethernet/phy/Kconfig.tja11xx:
```diff
@@ -0,0 +1,11 @@
+# SPDX-License-Identifier: Apache-2.0
+
+config PHY_TJA11xx
+	bool "TJA11xx PHY driver"
```
``depends on DT_HAS_NXP_TJA11XX_ENABLED`` missing

KevShaju replied:
Added. Thanks.

maass-hamburg commented on drivers/ethernet/phy/Kconfig.tja11xx:
```diff
@@ -0,0 +1,12 @@
+# Copyright 2024 NXP
+# SPDX-License-Identifier: Apache-2.0
+
+config PHY_TJA11xx
+	bool "TJA11xx PHY driver"
+	default y
+	depends on DT_HAS_NXP_TJA11xx_ENABLED
```
Kconfigs are all uppercase, use uppercase XX

maass-hamburg commented on drivers/ethernet/phy/phy_tja11xx.c:
```diff
@@ -0,0 +1,261 @@
+
+struct phy_tja11xx_config {
+	const struct device *mdio;
+	struct gpio_dt_spec gpio_interrupt;
```
remove `struct gpio_dt_spec gpio_interrupt;` it is not used

KevShaju replied:
removed

maass-hamburg commented on drivers/ethernet/phy/phy_tja11xx.c:
```diff
@@ -0,0 +1,261 @@
+	static const struct phy_tja11xx_config phy_tja11xx_config_##n = {                          \
+		.phy_addr = DT_INST_REG_ADDR(n),                                                   \
+		.mdio = DEVICE_DT_GET(DT_INST_BUS(n)),                                             \
+		.gpio_interrupt = GPIO_DT_SPEC_INST_GET_OR(n, int_gpios, {0}),                     \
```
here too

KevShaju replied:
removed

maass-hamburg commented on drivers/ethernet/phy/phy_tja11xx.c:
```diff
@@ -0,0 +1,261 @@
+	};                                                                                         \
+	static struct phy_tja11xx_data phy_tja11xx_data_##n = {                                    \
+		.sem = Z_SEM_INITIALIZER(phy_tja11xx_data_##n.sem, 1, 1),                          \
+		.offload_sem = Z_SEM_INITIALIZER(phy_tja11xx_data_##n.offload_sem, 0, 1),          \
```
unused

KevShaju replied:
removed

maass-hamburg commented on drivers/ethernet/phy/phy_tja11xx.c:
```diff
@@ -0,0 +1,261 @@
+	const struct device *dev;
+	struct phy_link_state state;
+	struct k_sem sem;
+	struct k_sem offload_sem;
```
unused

KevShaju replied:
removed

maass-hamburg reviewed (dismissed):
phy looks fine

marcin-wierzbicki commented:
> also as said by @manuargue I would also prefer it, that the eth phy driver got it's own PR.

The comments have been addressed, not sure if a separate PR is still necessary.

marcin-wierzbicki committed:
boards: nxp: add support for S32K148 evaluation board

Support for NXP S32K148 evaluation board (s32k148_evb).

Adapt samples: adc_dt, adc_sequence.
Adapt tests: adc_api, gpio_basic_api, gpio_hogs.

Signed-off-by: Marcin Wierzbicki <marcin.wierzbicki@accenture.com>

marcin-wierzbicki commented:
Updated the hal_nxp entry in the west manifest to point to nxp_hal PR.
Moved changes related to nxp_pedbg runner to separate commit.

Please review the PR.

KevShaju committed:
boards: shields: Add shield for NXP ADTJA1101 Ethernet Adapter

Add a shield for NXP ADTJA1101 Ethernet Adapter. This shield can
be used with the s32k148_evb.

Signed-off-by: Kevin Shaju <kevin.shaju@accenture.com>

KevShaju committed:
drivers: net: phy: Add tja11xx driver

Adds the c22 tja11xx driver.

Signed-off-by: Kevin Shaju <kevin.shaju@accenture.com>

kartben commented on boards/nxp/s32k148_evb/doc/index.rst:
```diff
@@ -0,0 +1,181 @@
+Hardware
+********
+
+- NXP S32K148
+    - Arm Cortex-M4F @ up to 112 Mhz
+    - 1.5 MB Flash
+    - 256 KB SRAM
+    - up to 127 I/Os
+    - 3x FlexCAN with FD
+    - eDMA, 12-bit ADC, MPU, ECC and more.
+
+- Interfaces:
+    - CAN, LIN, UART/SCI
+    - Ethernet connector compatible with different ethernet daughter cards
+    - 2 touchpads, potentiometer, user RGB LED and 2 buttons.
```
```suggestion
- NXP S32K148

  - Arm Cortex-M4F @ up to 112 Mhz
  - 1.5 MB Flash
  - 256 KB SRAM
  - up to 127 I/Os
  - 3x FlexCAN with FD
  - eDMA, 12-bit ADC, MPU, ECC and more.

- Interfaces

  - CAN, LIN, UART/SCI
  - Ethernet connector compatible with different ethernet daughter cards
  - 2 touchpads, potentiometer, user RGB LED and 2 buttons.
```

kartben commented on boards/nxp/s32k148_evb/doc/index.rst:
```diff
@@ -0,0 +1,185 @@
+   Before using the Ethernet interface, please take note of the following:
+
+   - For boards with the part number ``LSF24D`` at ``U16``, ``R553`` needs to be depopulated.
+
+.. note::
+   Before using the Ethernet interface, please take note of the following:
+   - For boards with the part number ``LSF24D`` at ``U16``, ``R553`` needs to be depopulated.
```
Duplicated content? Drop?

nordicjm commented on samples/net/sockets/dumb_http_server/CMakeLists.txt:
```diff
@@ -1,6 +1,9 @@
 # SPDX-License-Identifier: Apache-2.0
 
 cmake_minimum_required(VERSION 3.20.0)
+if("${BOARD}" STREQUAL "s32k148_evb/s32k148")
+  set(SHIELD nxp_adtja1101)
+endif()
```
don't do this, if it needs a shield, add it to the sample.yml file as a new entry which applies the shield

KevShaju replied:
Could you please provide an example of how this should be done?

kartben reviewed (dismissed):
+1 for docs

nordicjm commented on samples/posix/gettimeofday/sample.yaml:
```diff
@@ -13,3 +13,5 @@ common:
 tests:
   sample.posix.gettimeofday:
     harness: net
+    extra_args: SHIELD=nxp_adtja1101
```
none of these are right, you need to add a **new** entry to the tests which just runs on this board with this shield, what you have done here is prevent running the samples on any other board than your own

KevShaju replied:
Thank you for the explanation. Does it look right now?

manuargue commented on scripts/west_commands/runners/nxp_pedbg.py:
As far as I understand, you are proposing a west runner for P&E micro multilink debug probe to be used for NXP S32K1 devices through NXP S32DS IDE... If you plan to support P&E micro debug probe it would make more sense to create a generic runner for it that can work with any chip supported by the probe. Also note that P&E micro multilink debug probe is not from NXP.

If you want to enable a runner for this specific chip, I'd suggest for this first enablement PR to only enable JLink which supports this chip (https://www.segger.com/supported-devices/jlink/nxp/s32k1) and that there's already a west runner for it and move P&E micro multilink support to a separate PR. You can find a reference on how to enable JLink here: https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/nxp/ucans32k1sic/board.cmake#L4-L9

marcin-wierzbicki replied:
> As far as I understand, you are proposing a west runner for P&E micro multilink debug probe to be used for NXP S32K1 devices through NXP S32DS IDE... If you plan to support P&E micro debug probe it would make more sense to create a generic runner for it that can work with any chip supported by the probe. Also note that P&E micro multilink debug probe is not from NXP.
> 

Right. Let me implement that in a more generic way.

> If you want to enable a runner for this specific chip, I'd suggest for this first enablement PR to only enable JLink which supports this chip (https://www.segger.com/supported-devices/jlink/nxp/s32k1) and that there's already a west runner for it and move P&E micro multilink support to a separate PR. You can find a reference on how to enable JLink here: https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/nxp/ucans32k1sic/board.cmake#L4-L9

We tried jlink with these params:
```
board_runner_args(jlink
  "--device=S32K148"
  "--speed=4000"
  "--iface=jtag"
  "--reset"
)
```
Both, "west debug" and "west flash" are working. However, "west flash" waits for an input on stdin. After pressing Enter, it finishes about 1 second later and the board is flashed:
```
-- west flash: rebuilding
ninja: no work to do.
-- west flash: using runner jlink
-- runners.jlink: reset after flashing requested
-- runners.jlink: JLink version: 8.24
-- runners.jlink: Flashing file: ../zephyr/build/zephyr/zephyr.hex
```

manuargue replied:
I don’t have the hardware to test it unfortunately. You can see the JLink steps and examples available for K1, perhaps is required an additional cli option. 

For the PEmicro west runner, please do it in a separate PR because is not going to be specific to this device.

marcin-wierzbicki replied:
> You can try adding `board_runner_args(jlink "--tool-opt=-autoconnect 1")` or else run `west -v flash` and share the logs to have a look.

We confirmed that these params are correct - no more ENTER needs to be pressed:
```
board_runner_args(jlink
  "--device=S32K148"
  "--speed=4000"
  "--iface=jtag"
  "--reset"
  "--tool-opt=-jtagconf -1,-1"
)
```

> 
> For the PEmicro west runner, please do it in a separate PR because is not going to be specific to this device and this PR is getting already too large for a "adding support for NXP S32K148 evaluation board" :)

Of course.

nordicjm reviewed (dismissed):
Board/SoC parts OK

maass-hamburg commented:
@marcin-wierzbicki rebase needed, due to conflicts


@manuargue what is with hal change?

marcin-wierzbicki commented:
> @marcin-wierzbicki rebase needed, due to conflicts

nordicjm commented on boards/nxp/s32k148_evb/s32k148_evb.dts:
```diff
@@ -0,0 +1,182 @@
+			zephyr,code = <INPUT_KEY_1>;
+		};
+	};
+
```
```suggestion
```

nordicjm commented on dts/bindings/ethernet/nxp,tja11xx.yaml:
```diff
@@ -0,0 +1,23 @@
+    required: true
+    description: PHY address
+
+  master-slave:
+    type: string
+    description: |
+      100BASE-T1 Specifies that either phy has to run in master / slave mode
+      Default selects the mode set by the pinstrapping on the hardware design.
+    enum:
+      - "default"
+      - "master"
+      - "slave"
```
is there a replacement term for this? @kartben

kartben replied:
Thanks @nordicjm for bringing this up.

Please use "Leader" / "Follower" as recommend in IEEE 802.3-2022, Annex K. Please also update code accordingly (although it's actually not clear to me if the driver is -currently?- doing anything with this binding property).

KevShaju replied:
Changed the macros to `TJA11XX_...` as the TJA11xx PHYs have some common registers.

nordicjm replied:
unmarking as resolved because this is still here?

manuargue commented:
>@manuargue what is with hal change?

it's in discussion internally but is moving slow, I'll try to get back to you asap

manuargue commented:
@marcin-wierzbicki , pls rebase the hal_nxp branch and after this one. It seems that many unrelated twister failures may be fixed by rebasing, because these are not failing in current `main`. Besides there are some failures for the phy driver and board introduced on this pr that should be addressed

kartben reviewed (changes requested):
didn't realize this included a shield (would have been great to put in a different PR, fwiw) -- please add a shield.yml for it.
PR needs a rebase anyway

kartben commented on boards/nxp/s32k148_evb/s32k148_evb.dts:
```diff
@@ -0,0 +1,193 @@
+	leds {
+		compatible = "gpio-leds";
+
+		led1_red: led_0 {
+			gpios = <&gpioe 21 GPIO_ACTIVE_HIGH>;
+			label = "LED1_RGB_RED";
+		};
+
+		led1_green: led_1 {
+			gpios = <&gpioe 22 GPIO_ACTIVE_HIGH>;
+			label = "LED1_RGB_GREEN";
+		};
+
+		led1_blue: led_2 {
+			gpios = <&gpioe 23 GPIO_ACTIVE_HIGH>;
+			label = "LED1_RGB_BLUE";
+		};
```
I think you want `ACTIVE_LOW` for these LEDs?

![image](https://github.com/user-attachments/assets/d52648d0-e6f8-4680-86a6-a84c6c9b63d1)

kartben commented on boards/nxp/s32k148_evb/s32k148_evb.dts:
```diff
@@ -0,0 +1,193 @@
+	pwmleds {
+		compatible = "pwm-leds";
+
+		led1_red_pwm: led_pwm_0 {
+			pwms = <&ftm4 1 PWM_MSEC(20) PWM_POLARITY_NORMAL>;
+			label = "LED1_RGB_RED_PWM";
+		};
+
+		led1_green_pwm: led_pwm_1 {
+			pwms = <&ftm4 2 PWM_MSEC(20) PWM_POLARITY_NORMAL>;
+			label = "LED1_RGB_GREEN_PWM";
+		};
+
+		led1_blue_pwm: led_pwm_2 {
+			pwms = <&ftm4 3 PWM_MSEC(20) PWM_POLARITY_NORMAL>;
+			label = "LED1_RGB_BLUE_PWM";
+		};
```
same here, possibly want to have POLARITY_INVERTED

dkalowsk commented:
merged to unblock the CI failure
